GH-42240: [R] Fix crash in ParquetFileWriter$WriteTable and add WriteBatch

amoeba opened the pull request:
### Rationale for this change

See https://github.com/apache/arrow/issues/42240.

### What changes are included in this PR?

- Fixes a crash in `ParquetFileWriter$WriteTable` by asserting the class of what's passed in and stopping if it's not a `Table`
- Since I was already there, added `WriteBatch` to match [`pyarrow.parquet.ParquetWriter.write_batch`](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html#pyarrow.parquet.ParquetWriter.write_batch) which is just a convenience
- Adds a test for the behavior of trying to write to a closed sink
- Bumps the minimum Arrow C++ test version we test the R package with on CI from 13 to 15
- Removes one ARROW_VERSION_MAJOR >= 15 guard

### Are these changes tested?

Yes.

### Are there any user-facing changes?

New method on ParquetFileWriter (WriteBatch).
* GitHub Issue: #42240

jonkeane commented on r/R/parquet.R:
```diff
@@ -428,6 +428,12 @@ ParquetFileWriter <- R6Class("ParquetFileWriter",
+    },
+    WriteBatch = function(batch, chunk_size) {
+      assert_is(batch, "RecordBatch")
+      table <- Table$create(batch)
       parquet___arrow___FileWriter__WriteTable(self, table, chunk_size)
```
These methods are super close together, so I doubt they would drift, but just in case, would it make sense to call `WriteTable(table, ...)` here in case we add other assertions / changes to `WriteTable` and forget to copy them down here?

amoeba replied:
That makes plenty of sense, thanks for noticing. Fixed in 2a8fb0d9bb1113173d5eee7e7517b2b4234300c4.

jonkeane reviewed (approved):
Thanks for jumping on this! I have one comment, feel free to take it or leave it.

amoeba commented:
I pushed up a change addressing that comment and noticed I missed documenting the new class method so I added another commit for that. Letting CI run and then I'll merge.

amoeba commented:
One of the tests this PR adds, in conjunction with the minimum supported C++ version check, caught a segfault in Arrow C++ that wasn't fixed until Arrow 15. I think we need to put a closed-check in the R package somewhere to catch this. I'll look into that and include a note that the check can be removed once we bump the minimum version to or above 15.

amoeba commented:
It wasn't easy to expose the writer's `closed_` state to the R package for Arrow C++ 13/14 so I opted just to skip the test on Arrow C++ <15, see 22a706b2e2d5eaf7cbacbbcf6de6a9cfa2132a09. @jonkeane can you take a quick look?

jonkeane commented:
> It wasn't easy to expose the writer's closed_ state to the R package for Arrow C++ 13/14 so I opted just to skip the test on Arrow C++ <15, see https://github.com/apache/arrow/commit/22a706b2e2d5eaf7cbacbbcf6de6a9cfa2132a09. @jonkeane can you take a quick look?

Does this mean we should bump our minimum libarrow to 15 then? If we can't support it, that's the way to go IMO. The backwards compatibility with libarrow is nice, but as far as I know it's not used anywhere specifically. It was added with the hopes that we might use system-library libarrow on CRAN among other places, but that hasn't happened.

amoeba commented:
Hrm, maybe. I think the only thing to do other than that would be to duplicate the open/closed state tracking on the R side until we bump the minimum version for some other reason. I think it may be relevant that this segfault is just in the manual invocation of the ParquetFileWriter and I'm not sure if that impacts the decision-making process here.

jonkeane commented:
Do we know of any actual uses of the backwards compatibility in the wild?

amoeba commented:
I'm not aware of any.

jonkeane commented:
> I'm not aware of any.

In that case I think we should just require >15 and that's ok (that might also let us delete some of the `ifdefs` too).

thisisnic commented:
> Does this mean we should bump our minimum libarrow to 15 then?

Don't we need to be 13 for Windows or am I getting my number mixed up?

amoeba commented:
Okay @jonkeane, I've bumped the minimum version from 13 to 15 and updated the PR body. There was only one `ARROW_VERSION_MAJOR` guard to be removed which I removed.

jonkeane commented:
Thanks! Sorry I should have linked this before, but we also need to change https://github.com/apache/arrow/blob/89fd5664b942f0cec1c51a4a17610aac3015d080/r/tools/check-versions.R#L27 too yeah?

amoeba commented:
My fault for missing that. I changed that and added a NEWS.md entry while I was at it.

amoeba commented:
I think CI is happy now that I pushed https://github.com/apache/arrow/pull/42241/commits/c2d39d0dff6fa3f14389352b6436e94d7affa73c.
